### PR TITLE
IPYNB Refactor Merge into master

### PR DESCRIPTION
**Pull request by oscarhibbert**
Refactor of notebooks into main.py complete. Predict Model function handles polls, results, ONS and Google trends. See decoupling issue #57 in GitHub issue tracker for further info on specifying specific data source combinations

### CODE REVIEW

File: election_predictor/interface/main.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+# Import all functions inside data.py from ml_logic
+from election_predictor.ml_logic.data import fetch_clean_data
+
+# Import preprocessor
+from election_predictor.ml_logic.preprocessor import preprocessor
+
+# Import modelling functions from ml_logic
+from election_predictor.ml_logic.model import XGBoostModel
+
+def predict_election() -> dict:
```
Comment (nieksonneveld) [inline election_predictor/interface/main.py:17]: should this take election_year as an argument?

File: election_predictor/interface/main.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        # Handle XGBoost Regressor modelling
+        # Handle model fetch and initialisation
+        xgb_regressor = XGBoostModel()
+        xgbr_model = xgb_regressor.initialize_model()
+
+        # Handle model compiling
+        xgb_regressor.compile_model(xgbr_model) # Uses default model parameters from params.py
+
+        # Handle model training
+        trained_model = xgb_regressor.train_model(xgbr_model, X_train_matrix, party_y_train)
```
Comment (nieksonneveld) [inline election_predictor/interface/main.py:218]: does this instantiate a separate model for each party?

File: election_predictor/interface/main.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    election_predictions = { }
+
+    for party_code, model in trained_models.items():
+        trained_model = model
+
+        predicted_vote_share = trained_model.predict(X_predict_matrix).mean()
+
+        election_predictions[party_code] = predicted_vote_share
+
+    return election_predictions
```
Comment (nieksonneveld) [inline election_predictor/interface/main.py:243]: this currently only returns national vote share (without SCO correction?), where do constituency seats get predicted?

**Review by nieksonneveld — APPROVED** [review]

File: election_predictor/interface/main.py (hunk 1 above)
Comment (oscarhibbert) [reply election_predictor/interface/main.py:17]: For now the function is designed to predict the 2024 election only. Next step will be for it to take previous election years also as the UI will need to fetch results for these years (if we stick to current design). My aim is to have live data cleaned, fetched, and predicted on for 2024 as a sort of MVP and then go from there.

Additionally, logic for evaluating model performance against previous election results is yet to be added but should be to optimise for the most performant model as live data sourcing could alter performance.

File: election_predictor/interface/main.py (hunk 2 above)
Comment (oscarhibbert) [reply election_predictor/interface/main.py:218]: Yes

File: election_predictor/interface/main.py (hunk 3 above)
Comment (oscarhibbert) [reply election_predictor/interface/main.py:243]: Constituency seats logic that CK built is coming to the function soon, for initial MVP of live source data am focussing on vote share only and will add this in the coming days.

**Pull request merged**